rasa_nlu_examples/tokenizers/stanzatokenizer.py

```diff
@@ -0,0 +1,128 @@
+from rasa.nlu.training_data import Message
+from functools import reduce
+from rasa.nlu.tokenizers.spacy_tokenizer import POS_TAG_KEY
+
+import stanza
+
+
+class StanzaTokenizer(Tokenizer):
+    """
+    The StanzaTokenizer allows for more pos/lemma features to be used in the
+    Rasa ML pipelines. It is based on the project found here: https://stanfordnlp.github.io/stanza/usage.html
+
+    Not every language here has good performance metrics. To see the details
+    check out this table: https://stanfordnlp.github.io/stanza/performance.html
+
+    Before running the stanza model in production, be sure to check the license information
+    since it may differ per language: https://stanfordnlp.github.io/stanza/available_models.html
+    """
+
+    defaults = {
+        # What language to use
+        "lang": None,
+        # What package to use, details on this can be found here:
+        # https://stanfordnlp.github.io/stanza/download_models.html
+        "pkg": "package",
+        # Where to load the model
+        "cache_dir": None,
+    }
+
+    # the following language should not be tokenized using the WhitespaceTokenizer
+    language_list = [
+        "af",
+        "grc",
+        "ar",
+        "hy",
+        "eu",
+        "be",
+        "bg",
+        "bxr",
+        "ca",
+        "zh",
+        "lzh",
+        "cop",
+        "hr",
+        "cs",
+        "da",
+        "nl",
+        "en",
+        "et",
+        "fi",
+        "fr",
+        "gl",
+        "de",
+        "got",
+        "el",
+        "he",
+        "hi",
+        "hu",
+        "id",
+        "ga",
+        "it",
+        "ja",
+        "ko",
+        "kmr",
+        "lv",
+        "lt",
+        "olo",
+        "mt",
+        "sme",
+        "no",
+        "nn",
+        "cu",
+        "fro",
+        "orv",
+        "fa",
+        "pl",
+        "pt",
+        "ro",
+        "ru",
+        "gd",
+        "sr",
+        "sk",
+        "sl",
+        "es",
+        "sv",
+        "swl",
+        "ta",
+        "te",
+        "tr",
+        "uk",
+        "hsb",
+        "ur",
+        "ug",
+        "vi",
+        "wo",
+    ]
+
+    def __init__(self, component_config: Dict[Text, Any] = None) -> None:
+        """Construct a new tokenizer using the Stanza framework."""
+
+        super().__init__(component_config)
+        self.nlp = stanza.Pipeline(
+            lang=component_config["lang"],
+            dir=component_config["cache_dir"],
+            processors="tokenize,pos,lemma",
+            tokenize_no_ssplit=True,
+        )
+
+    def tokenize(self, message: Message, attribute: Text) -> List[Token]:
+        text = message.get(attribute)
+
+        doc = self.nlp(text)
+        stanza_tokens = reduce(lambda a, b: a + b, doc.sentences).tokens
+        tokens = [
+            Token(
+                text=t.text,
+                start=t.start_char,
+                end=t.end_char,
+                lemma=t.words[0].lemma,
```

Comment:
No idea how stanza works, but why can there be multiple words for a token? Or is there always exactly one word per token?

Reply:
My impression is that stanza, for some languages, also offers [multi word tokens](https://stanfordnlp.github.io/stanza/mwt.html). My assumption was that it was related to this.

Reply:
But in that case lemma and text would not match, would they? Maybe we should add a check and no fill the lemma in case they are multiple tokens?

Reply:
(same for POS tags)